pages/en/lb3/Discovering-models-from-relational-databases.md

```diff
@@ -23,69 +23,182 @@ This process is called _discovery_ and is supported by the following connectors:
+      for (const modelName in models) {
+        app.model(models[modelName], {dataSource: dataSource});
+      }
+
+      callback();
     });
-  });
-});
+};
+```
+
+Next steps:
+
+ 1. Start your application in the usual way, e.g. via `npm start` or `node .`
+ 2. Use [API Explorer](./Use-API-Explorer.md) to inspect your dynamically
+    created REST API.
+
+## Discover and save model definitions
+
+Use [dataSource.discoverSchema()](http://apidocs.loopback.io/loopback-datasource-juggler/#datasource-prototype-discoverschema)
+API to obtain a model definition for a given database table.
+
+To leverage this API, you need to write a small script to discover the models
+you are interested in and write them to model JSON files.
+
+Let's walk through an example implementation of such script.
+
+{% include note.html content="
+The code examples in this section are using features introduced in
+Node.js 8.x LTS. Users running on an older version of Node.js need
+to rework the code to use promises (or callbacks) instead of async/await
+for flow control.
+" %}
+
+First of all, we need to import APIs we will be using later:
+
+{% include code-caption.html content="/discover-models.js" %}
+```js
+'use strict';
+
+const loopback = require('loopback');
+const promisify = require('util').promisify;
+const fs = require('fs');
+const writeFile = promisify(fs.writeFile);
+const readFile = promisify(fs.readFile);
+const mkdirp = promisify(require('mkdirp'));
+```
+
+Secondly, we need to initialize the data source backed by the database
+we want to discover our models from. While it's possible to load the full
+application and let loopback-boot load all artifacts (including data sources),
+we recommend a simpler and faster approach of creating the data source
+instance explicitly:
+
+```js
+const DATASOURCE_NAME = 'db';
+const dataSourceConfig = require('./server/datasources.json');
+const db = new loopback.DataSource(dataSourceConfig[DATASOURCE_NAME]);
 ```
 
+Node.js does not allow the main module function to be async,
+the top-level code cannot use `await` keyword for flow control.
+As a workaround, we need to wrap the discovery code in a helper function
+that's executed using Promise API.
+
+```js
+discover().then(
+  success => process.exit(),
+  error => { console.error('UNHANDLED ERROR:\n', error); process.exit(1); },
+);
+```
+
+With the preparations done, it's time to roll up our sleeves
+and do the actual model discovery.
+
+```js
+async function discover() {
+  // It's important to pass the same "options" object to all calls
+  // of dataSource.discoverSchemas(), it allows the method to cache
+  // discovered related models
+  const options = {relations: true};
+
+  // Discover models and relations
+  const salariesSchemas = await db.discoverSchemas('salaries', options);
+  const employeesSchemas = await db.discoverSchemas('employees', options);
+
+  // Create model definition files
+  await mkdirp('common/models');
```

Comment:
I'd personally find using `mkdirSync` more readable. https://nodejs.org/api/fs.html#fs_fs_mkdirsync_path_mode

Reply:
I need `mkdirp` because the directory `common` may not exist yet. It has a sync API too ([docs](https://www.npmjs.com/package/mkdirp#mkdirpsyncdir-opts)). I prefer `await` over sync API, I find it more idiomatic.

Or are you suggesting to get rid of `mkdirp` and call `mkdir`/`mkdirSync` twice, once for each directory? I find that cumbersome.

Reply:
> I find it more idiomatic.